torch/csrc/jit/script/compiler.cpp

```diff
@@ -830,23 +939,67 @@ struct to_ir {
   }
 
   Value* emitTernaryIf(const TernaryIf& expr) {
+    auto bool_info = findRefinements(expr.cond());
     Value* cond_value = emitCond(expr.cond());
-    auto true_expr = [&] { return emitExpr(expr.true_expr()); };
-    auto false_expr = [&] { return emitExpr(expr.false_expr()); };
+    auto true_expr = [&] {
+      insertRefinements(bool_info.true_refinements);
+      return emitExpr(expr.true_expr());
+    };
+    auto false_expr = [&] {
+      insertRefinements(bool_info.false_refinements);
+      return emitExpr(expr.false_expr());
+    };
     return emitIfExpr(expr.range(), cond_value, true_expr, false_expr);
   }
 
+  void insertRefinements(const Refinements& ref) {
+    const auto& types = ref.types;
+    const auto& ranges = ref.ranges;
+
+    for (auto value_type: types) {
+      const std::string& name = value_type.first;
+      auto type = value_type.second;
+      auto range = *ranges.at(name);
+      Value * v = environment_stack->getVar(name, range);
+      if (type != NoneType::get()) {
+        auto output = graph->insert(prim::unchecked_unwrap_optional, {v});
+        environment_stack->setVar(range, name, output);
+      }
+    }
+  }
+
   Value* emitShortCircuitIf(
       const SourceRange& loc,
       const TreeRef& first_expr,
       const TreeRef& second_expr,
       bool is_or) {
+
+    auto first_bool_info = findRefinements(first_expr);
     Value* first_value = emitCond(Expr(first_expr));
 
-    auto get_first_expr = [first_value] { return first_value; };
-    auto get_second_expr = [&] { return emitCond(Expr(second_expr)); };
+    const Refinements* first_expr_refinements;
+    const Refinements* second_expr_refinements;
+    // if it's an OR the first expr is emitted in the true branch
+    // and the second expr in the false branch, if it's an AND the opposite
+    if (is_or) {
+      first_expr_refinements = &first_bool_info.true_refinements;
+      second_expr_refinements = &first_bool_info.false_refinements;
+    } else {
+      first_expr_refinements = &first_bool_info.false_refinements;
+      second_expr_refinements = &first_bool_info.true_refinements;
+    }
+
+    auto get_first_expr = [&] {
+      insertRefinements(*first_expr_refinements);
```

Comment:
Refinements here are not scoped appropriately. They will stay life after the end of the condition and through the end of the block the conditional is in.

Reply:
Could you elaborate ?

Reply:
These expressions are emitted when each corresponding block is emitted so I don't think they escape the condition. I will add a test to make sure this does not regress.